### ISSUE TITLE
Requests made during no network connectivity results in non cached results and fails to be retrieved after mutation is performed

### ISSUE THREAD

**Issue by lawmicha** [awslabs/aws-mobile-appsync-sdk-ios#390]
Is this issue still being investigated? I just encountered this as I was creating a light example app locally.

The repro path is pretty straight forward, I can include my example code here (mostly pulled directly from the integration docs):

In AppDelegate.swift

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        do {
            // You can choose the directory in which AppSync stores its persistent cache databases
            let cacheConfiguration = try AWSAppSyncCacheConfiguration()
            
            // AppSync configuration & client initialization
            let appSyncServiceConfig = try AWSAppSyncServiceConfig()
            let appSyncConfig = try AWSAppSyncClientConfiguration(appSyncServiceConfig: appSyncServiceConfig,
                                                                  cacheConfiguration: cacheConfiguration)
            self.appSyncClient = try AWSAppSyncClient(appSyncConfig: appSyncConfig)
            // Set id as the cache key for objects. See architecture section for details
            self.appSyncClient?.apolloClient?.cacheKeyForObject = { $0["id"] }
        } catch {
            print("Error initializing appsync client. \(error)")
        }
        // other methods
        return true
    }

In ViewController.swift

    //Reference AppSync client
    var appSyncClient: AWSAppSyncClient?
    {
        get
        {
            return (UIApplication.shared.delegate as! AppDelegate).appSyncClient
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        let button = UIButton.init(type: .roundedRect)
        button.setTitle("Create Item", for: .normal)
        button.addTarget(self, action: #selector(createItem), for: .touchUpInside)
        button.frame = .init(x: 25, y: 250, width: 200, height: 100)
        
        let findButton = UIButton.init(type: .roundedRect)
        findButton.setTitle("List Items (cached)", for: .normal)
        findButton.addTarget(self, action: #selector(listItemsCached), for: .touchUpInside)
        findButton.frame = .init(x: 25, y: 450, width: 200, height: 100)
        
        self.view.addSubview(button)
        self.view.addSubview(findButton)
    }
    
    @objc
    func createItem()
    {
        let tempID = "temp"+UUID.init().uuidString
        let value = "I have a value of '" + tempID + "'"
        let mutation = CreateItemMutation(id: tempID, value: value)
        self.appSyncClient?.perform(mutation: mutation, optimisticUpdate: { (transaction) in
            do {
                try transaction?.update(query: GetItemsQuery()) { (data: inout GetItemsQuery.Data) in
                    print("Handling optimistic update.")
                    data.getItems?.append(GetItemsQuery.Data.GetItem.init(id: tempID, value: value))
                }
            } catch {
                print("Error updating cache with optimistic response")
            }
        }, resultHandler: { (result, error) in
            if let result = result {
                print("Added Response from service: \(String(describing: result.data?.createItem?.value))")
                //Now remove the outdated entry in cache from optimistic write
                let _ = self.appSyncClient?.store?.withinReadWriteTransaction { transaction in
                    try transaction.update(query: GetItemsQuery())
                    { (data: inout GetItemsQuery.Data) in
                        var pos = -1, counter = 0
                        for item in (data.getItems!) {
                            if item?.id == tempID {
                                pos = counter
                                continue
                            }; counter += 1
                        }
                        if pos != -1 { data.getItems?.remove(at: pos) }
                    }
                }
            } else if let error = error {
                print("Error adding Item: \(error.localizedDescription)")
            }
        })
    }
    
    @objc
    func listItemsCached()
    {
        appSyncClient?.fetch(query: GetItemsQuery(), cachePolicy: .returnCacheDataAndFetch) {
            (result, error) in
            if error != nil {
                print(error?.localizedDescription ?? "")
                return
            }
            print("Results:")
            result?.data?.getItems?.forEach { print($0!.value) }
        }
    }

If you set up a simple iOS app like this, launch it *for the first time* with internet connectivity disabled, press the "Create Item" button, and then press the "List Items (cached)" button, you'll get no results. 

However, if you open the application with internet, press "List Items (cached)", turn off the internet, and then press "Create Item" then subsequent presses of "List Items (cached)" will return results of the optimistic update. I've attached two gifs below showing the difference in behavior with the example code I've provided:

Example with no internet connection for duration of demonstration: Note that the code for `listItemsCached` is just returning an error, no cached results, even after a mutation is performed.
![tmp_no_internet](https://user-images.githubusercontent.com/1887444/56188197-b2928700-5fd9-11e9-84b4-6deb34f11d60.gif)

This below example has internet turned on ONLY for the first press of "List Items (Cached)", then I turned it off and you can observe the expected behavior.
![tmp_internet_for_first_cache](https://user-images.githubusercontent.com/1887444/56188506-8c211b80-5fda-11e9-8dfd-396dcd089e03.gif)

Happy to help provide any additional info if needed!

_Originally posted by @camerona93 in https://github.com/awslabs/aws-mobile-appsync-sdk-ios/issues/92#issuecomment-483536608_

**Comment by royjit** [comment]
Adding some investigation done so far:

schema used

```json
type Todo @model {
  id: ID!
  name: String!
  description: String
}

```
The queries.db values are different in the two scenarios. The db file can be found at the path - /Library/Caches/appsync/queries.db 

**When initially offline**

1.  Turn off wifi
2. Launch the app
3. Do a query to list items
4. Add an item
5. Check the db table `records`


_id|key|record
-- | -- | --
2 | QUERY_ROOT | {"listTodos":null}


**When initially online**
1.  Turn on wifi
2. Launch the app
3. Do a query to list items
4. Turn off wifi
5. Add an item
6. Check the db table `records`

_id | key | record
-- | -- | --
2 | QUERY_ROOT | {"listTodos":{"$reference":"QUERY_ROOT.listTodos"}}
3 | QUERY_ROOT.listTodos | {"__typename":"ModelTodoConnection","nextToken":null,"items":[]}

Wondering whether this is causing the query to return empty. Need to debug further inside the working of Apollo SDK.

**Comment by palpatim** [comment]
As demonstrated in the `doMutationWithOptimisticUpdate` Unit test, this behavior works as expected in AppSync. The problem is that the optimistic update block is doing option chaining to append the results to the transaction data block, rather than creating a new result if the data results are nil. 

tl;dr:

Change
```swift
data.getItems?.append(GetItemsQuery.Data.GetItem.init(id: tempID, value: value))
```

to

```swift
data.getItems = data.getItems ?? []
data.getItems?.append(GetItemsQuery.Data.GetItem.init(id: tempID, value: value))
```

**Comment by royjit** [comment]
PR to fix the documentation code - https://github.com/aws-amplify/docs/pull/3050
Closing this as there are no other action item with the Amplify team.
